internal/locate/region_cache.go

```diff
@@ -1666,12 +1666,18 @@ func (c *RegionCache) OnRegionEpochNotMatch(bo *retry.Backoffer, ctx *RPCContext
 	newRegions := make([]*Region, 0, len(currentRegions))
 	// If the region epoch is not ahead of TiKV's, replace region meta in region cache.
 	for _, meta := range currentRegions {
-		if _, ok := c.pdClient.(*CodecPDClient); ok {
-			var err error
+		var err error
+		oldMeta := meta
+		switch c.pdClient.(type) {
```

Comment:
Why not move `decodeRegionMetaKeyWithShallowCopy` into interface and use interface to dispatch.

Reply:
Because c.pdClient is a `pd.Client` interface already, if we change the interface, we need to modify the PD.